### PR TITLE
Adds support for recording. Closes #93

### LINKED ISSUE #93: Allow recording requests
Users should be able to put the Proxy in recording mode where it will capture requests. After recording is stopped, the Proxy will be able to execute an action on the recorded requests, like analyzing it for minimal permissions, building a report, etc.

Depends on #58

### PR DESCRIPTION
**Pull request by waldekmastykarz**
Adds support for recording. Closes #93

### CODE REVIEW

File: msgraph-developer-proxy/ProxyEngine.cs [hunk 1]
```diff
@@ -95,11 +98,80 @@ public class ProxyEngine {
...
+        var clonedLogs = _requestLogs.ToArray();
+        _requestLogs.Clear();
+        _pluginEvents.RaiseRecordingStopped(new RecordingArgs(clonedLogs));
+    }
+
+    private void PrintRecordingIndicator() {
+        lock (ConsoleLogger.ConsoleLock) {
+            if (_isRecording) {
+                Console.ForegroundColor = ConsoleColor.Red;
+                Console.Error.Write("◉");
```
Comment (waldekmastykarz) [inline msgraph-developer-proxy/ProxyEngine.cs:165]: I looked into making the indicator fixed in the terminal window but it turned out to be too complex for the current setup with continuous logging, without using any TUI framework. For now, the indicator is logged one-time to stderr.

File: msgraph-developer-proxy/ProxyEngine.cs [hunk 2]
```diff
@@ -95,11 +98,80 @@ public class ProxyEngine {
...
+        // clone the list so that we can clear the original
+        // list in case a new recording is started, and
+        // we let plugins handle previously recorded requests
+        var clonedLogs = _requestLogs.ToArray();
+        _requestLogs.Clear();
+        _pluginEvents.RaiseRecordingStopped(new RecordingArgs(clonedLogs));
+    }
+
+    private void PrintRecordingIndicator() {
+        lock (ConsoleLogger.ConsoleLock) {
```
Comment (waldekmastykarz) [inline msgraph-developer-proxy/ProxyEngine.cs:162]: I use a lock here to ensure that other log messages won't get mixed

**Review by gavinbarron — COMMENTED** [review]
One question on the indicator printed on stopping of recording, this looks solid though

File: msgraph-developer-proxy/ProxyEngine.cs [hunk 3]
```diff
@@ -95,11 +98,80 @@ public class ProxyEngine {
...
+        PrintRecordingIndicator();
+    }
+
+    private void StopRecording() {
+        if (!_isRecording) {
+            return;
+        }
+
+        _isRecording = false;
+        PrintRecordingIndicator();
```
[the remark this answers is not in the document]
Comment (gavinbarron) [reply msgraph-developer-proxy/ProxyEngine.cs:152]: That works for me.

**Pull request merged**